meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java

```diff
@@ -0,0 +1,174 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.concurrent.TimeUnit;
+
+public class DistributedSemaphore {
+  private final MetaClientInterface<DataRecord> _metaClient;
+  private final String _path;
+  private static final String INITIAL_CAPACITY_NAME = "INITIAL_CAPACITY";
+  private static final String REMAINING_CAPACITY_NAME = "REMAINING_CAPACITY";
+  private static final long DEFAULT_REMAINING_CAPACITY = -1;
+  private static final Logger LOG = LoggerFactory.getLogger(DistributedSemaphore.class);
+
+  public DistributedSemaphore(MetaClientConfig config, String path, int capacity) {
+    if (MetaClientConfig.StoreType.ZOOKEEPER.equals(config.getStoreType())) {
+      ZkMetaClientConfig zkMetaClientConfig = new ZkMetaClientConfig.ZkMetaClientConfigBuilder()
+          .setConnectionAddress(config.getConnectionAddress())
+          .setZkSerializer(new DataRecordSerializer()) // Currently only support ZNRecordSerializer.
+          // Setting DataRecordSerializer as DataRecord extends ZNRecord.
+          .build();
+      _metaClient = new ZkMetaClientFactory().getMetaClient(zkMetaClientConfig);
+      _metaClient.connect();
+    } else {
+      throw new MetaClientException("Unsupported store type: " + config.getStoreType());
+    }
+
+    _path = path;
+    if (_metaClient.exists(path) == null) {
+      DataRecord dataRecord = new DataRecord(path);
+      dataRecord.setLongField(INITIAL_CAPACITY_NAME, capacity);
+      dataRecord.setLongField(REMAINING_CAPACITY_NAME, capacity);
+      _metaClient.create(path, dataRecord);
+    }
+  }
+
+  public DistributedSemaphore(MetaClientInterface<DataRecord> client, String path) {
+    _metaClient = client;
+    _path = path;
+    _metaClient.connect();
+  }
+
+  /**
+   * Acquire a permit. If no permit is available, block until when it was able to acquire.
+   * @return a permit
+   */
+  public Permit acquire() {
+    if (getRemainingCapacity() > 0) {
+      updateAcquirePermit(1);
+      return retrievePermit(_path);
+    } else {
+      throw new MetaClientException("No sufficient permits available");
+    }
+  }
+
+
+  /**
+   * Try to acquire a permit. If not enough permit is available, wait for a specific time or return when it was able to acquire.
+   * @param count number of permits to acquire
+   * @return a collection of permits
+   */
+  public Collection<Permit> acquire(int count) {
+    if (getRemainingCapacity() < count) {
+      throw new MetaClientException("No sufficient permits available");
+    } else {
+      updateAcquirePermit(count);
+      Collection<Permit> permits = new ArrayList<>();
+      for (int i = 0; i < count; i++) {
+        permits.add(retrievePermit(_path));
+      }
+      return permits;
+    }
+  }
+
+  /**
+   * Try to acquire a permit. If no enough permit is available, wait for a specific time or return when it was able to acquire.
```

Comment:
tryAcquire should not be a blocking call -> meaning we dont wait but just return directly.

Reply:
Oh i see, then I must have gotten both acquire and tryAcquire mixed up. Should acquire then be blocking?